File: prac08/privateMethods.rb
class Example
  def public_method
    puts "This is a public method"
    private_method
  end

  private

  def private_method
    puts "This is a private method"
  end

  private
  attr_reader :private_attribute

  def initialize
    @private_attribute = "This is a private attribute"
  end
end

example = Example.new
example.public_method  # Works    
# example.private_method  # Raises NoMethodError
# puts example.private_attribute  # Raises NoMethodError

File: prac08/privateMethods.py
class Example:
    def __init__(self):
        self.__private_attribute = "This is a 'private' attribute"

    def public_method(self):
        print("This is a public method")
        self.__private_method()

    def __private_method(self):
        print("This is a 'private' method")

example = Example()
example.public_method()  # Works
# example.__private_method()  # Raises AttributeError
# print(example.__private_attribute)  # Raises AttributeError


# However, name mangling can be bypassed:
example._Example__private_method()  # Works, but not recommended: name mangling is intended to be a way to prevent accidental access to private attributes and methods
print(example._Example__private_attribute)  # Works, but not recommended

